src/SymmetricPositiveDefinite.jl

```diff
@@ -0,0 +1,272 @@
+
+The manifold of symmetric positive definite matrices, i.e.
+
+```math
+\mathcal P(n) =
+\bigl\{
+    x \in \mathbb R^{n\\times n} :
+    \xi^\mathrm{T}x\xi > 0 \text{ for all } \xi \in \mathbb R^{n}\backslash\{0\}
+\bigr\}
+```
+
+# Constructor
+
+    SymmetricPositiveDefinite(n)
+
+generates the $\mathcal P(n) \subset \mathbb R^{n\times n}$
+"""
+struct SymmetricPositiveDefinite{N} <: Manifold end
+SymmetricPositiveDefinite(n::Int) = SymmetricPositiveDefinite{n}()
+
+@doc doc"""
+    manifold_dimension(::SymmetricPositiveDefinite{N})
+
+returns the dimension of the manifold [`SymmetricPositiveDefinite`](@ref) $\mathcal P(n)$, N\in \mathbb N$, i.e.
+```math
+    \frac{n(n+1)}{2}    
+```
+"""
+@generated manifold_dimension(::SymmetricPositiveDefinite{N}) where {N} = div(N*(N+1), 2)
+
+@doc doc"""
+    LinearAffineMetric <: Metric
+
+The linear affine metric is the metric for symmetric positive definite matrices, that employs
+matrix logarithms and exponentials, which yields a linear and affine metric.
+"""
+struct LinearAffineMetric <: Metric end
+
+@doc doc"""
+    LogEuclideanMetric <: Metric
+
+The LogEuclidean Metric consists of the Euclidean metric applied to all elements after mapping them
+into the Lie Algebra, i.e. performing a matrix logarithm beforehand.
+"""
+struct LogEuclideanMetric <: Metric end
+
+@doc doc"""
+    distance(P,x,y)
+
+computes the distance on the [`SymmetricPositiveDefinite`](@ref) manifold between `x` and `y`,
+which defaults to the [`LinearAffineMetric`](@ref) induces distance.
+"""
+distance(P::SymmetricPositiveDefinite{N},x,y) where N = distance(MetricManifold(P,LinearAffineMetric()),x,y)
+@doc doc"""
+    distance(lP,x,y)
+
+computes the distance on the [`SymmetricPositiveDefinite`](@ref) manifold between `x` and `y`,
+as a [`MetricManifold`](@ref) with [`LinearAffineMetric`](@ref). The formula reads
+
+```math
+d_{\mathcal P(n)}(x,y) = \lVert \operatorname{Log}(x^{-\frac{1}{2}}yx^{-\frac{1}{2}})\rVert.
+```
+"""
+function distance(P::MetricManifold{SymmetricPositiveDefinite{N},LinearAffineMetric},x,y) where N
+    s = real.( eigen( x,y ).values )
+    return any(s .<= eps() ) ? 0 : sqrt(  sum( abs.(log.(s)).^2 )  )
+end
+
+@doc doc"""
+    inner(P,x,v,w)
+
+compute the inner product of `v`, `w` in the tangent space of `x` on the [`SymmetricPositiveDefinite`](@ref)
+manifold `P`, which defaults to the [`LinearAffineMetric`](@ref).
+"""
+inner(P::SymmetricPositiveDefinite{N}, x, w, v) where N = inner(MetricManifold(P,LinearAffineMetric()),x,w,v)
+@doc doc"""
+    inner(P,x,v,w)
+
+compute the inner product of `v`, `w` in the tangent space of `x` on the [`SymmetricPositiveDefinite`](@ref)
+manifold `P`, as a [`MetricManifold`](@ref) with [`LinearAffineMetric`](@ref). The formula reads
+
+```math
+( v, w)_x = \operatorname{tr}(x^{-1}\xi x^{-1}\nu ),
+```
+"""
+function inner(P::MetricManifold{SymmetricPositiveDefinite{N}, LinearAffineMetric}, x, w, v) where N
+    F = factorize(x)
```

Comment:
I'd actually use `F = chol(Symmetric(x))` here. This asserts that `x` is positive definite, and will throw if it's not. I'm not sure it makes a difference, but you may want to consider also wrapping `w` and `v` by Symmetric. If they are already of `Symmetric` type, this is going to be a no-op, but otherwise will help the compiler choose the right paths to do linear algebra.

Reply:
Interesting! Wasn't it you who proposed factorize? I am not yet that familiar with these optimizations, so I am happy to learn :)

Reply:
It was a past me. 😃 If all went well, `factorize` would choose `chol` anyway after checking symmetry and positive-definiteness. It just occurred to me that explicitly calling `chol` would immediately throw if any of our assumptions on the structure of `x` were not satisfied, but otherwise would go straight to what `factorize` would be supposed to do.

Reply:
Hm, but you actually mean `cholesky`? I can't find `chol` in `LinearAlgebra`...

Reply:
Yes, sorry, it used to be `cholfact` some time ago, and I vaguely remembered that they removed the `fact` from all factorizations, but not what else they did. 😄

Reply:
Now I am a little confused, because in 1.1 I even can't find `cholfact` but only `cholesky` (which is the only thing I found) does not work at that place. Any other ideas?

Reply:
Why does `cholesky` not work? ("some time ago" was referring to pre-v1.0 times, I guess). Of what type is `x`? A matrix, or a vector?

Reply:
`x` is a symmetric positive definite matrix. The old code works fine (`F = factorize(x)`), changing it so `F = factorize(Symmetric(x))` yields

>  no method matching /(::Symmetric{Float64,Array{Float64,2}}, ::BunchKaufman{Float64,Array{Float64,2}})

(yes I switched `w` and `v` to `Symmetric(w)` and `Symmetric(v)`),

and for `cholesky(Symmetric(x))` we get the same

>   MethodError: no method matching /(::Symmetric{Float64,Array{Float64,2}}, ::Cholesky{Float64,Array{Float64,2}})

and even without the `Symmetric` in there we get the same

>   MethodError: no method matching /(::Symmetric{Float64,Array{Float64,2}}, ::Cholesky{Float64,Array{Float64,2}})

And since I was trying anyways the variant

```
    F = cholesky(x)
    return tr( ( w / F ) * ( v / F ) )
```

also yields

> no method matching /(::Array{Float64,2}, ::Cholesky{Float64,Array{Float64,2}})

So in total: your idea of just `cholesky` does not seem to fit the linear system solving function afterwards.

Reply:
Ok, I'm very sorry. Then use whatever works for now, and maybe leave a comment for later. There seems to be no right solving for cholesky factorizations, which could/should be fixed then. I wonder what other factorization the generic `factorize` picks, probably `lu`... Meanwhile, I'll keep this in mind and see what can be done in `LinearAlgebra`.

Reply:
Nevermind, I learned a few things in `LinearAlgebra` at least and was anyways finding the last few bugs in the code :)